yt/frontends/stream/data_structures.py

```diff
@@ -1092,11 +1138,73 @@ def __init__(self, stream_handler, storage_filename=None,
         if ('io', 'density') in fields and ('io', 'smoothing_length') in fields:
             self._sph_ptype = 'io'
 
-def load_particles(data, length_unit = None, bbox=None,
-                   sim_time=0.0, mass_unit = None, time_unit = None,
+    def add_sph_fields(self, n_neighbors=32, kernel="cubic", sph_ptype="io"):
+        """Add SPH fields for the specified particle type.
+
+        For a particle type with "particle_position" and "particle_mass" already
+        defined, this method adds the "smoothing_length" and "density" fields.
+        "smoothing_length" is computed as the distance to the nth nearest
+        neighbor. "density" is computed as the SPH (gather) smoothed mass.
+
+        Caution: this method will overwrite existing "smoothing_length" and
+        "density" fields.
+
+        Parameters
+        ----------
+        n_neighbors : int
+            The number of neighbors to use in smoothing length computation.
+        kernel : str
+            The kernel function to use in density estimation.
+        sph_ptype : str
+            The SPH particle type. Each dataset has one sph_ptype only. This
+            method will overwrite existing sph_ptype of the dataset.
+
+        """
+        mylog.info("Generating SPH fields")
+
+        # Unify units
+        l_unit = "code_length"
+        m_unit = "code_mass"
+        d_unit = "code_mass / code_length**3"
+
+        # Read basic fields
+        ad = self.all_data()
+        pos = ad[sph_ptype, "particle_position"].to(l_unit).d
+        mass = ad[sph_ptype, "particle_mass"].to(m_unit).d
+
+        # Construct k-d tree
+        left_edge = self.domain_left_edge.to(l_unit).v
```

Comment:
Super minor nit here, but `YTArray` has a `to_value` method which basically does the same as `to(l_unit).v` does here.

Reply:
Thanks! I don't know that before.